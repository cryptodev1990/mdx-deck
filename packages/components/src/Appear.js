import React from 'react'
import Steps from './Steps'
import useSteps from './useSteps'

export const _Appear = props => {
  const arr = React.Children.toArray(props.children)
  return (
    <Steps
      length={arr.length}
      render={({ step }) => {
        const children = arr.map((child, i) =>
          i < step
            ? child
            : React.cloneElement(child, {
                style: {
                  ...child.props.style,
                  visibility: 'hidden',
                },
              })
        )
        return <>{children}</>
      }}
    />
  )
}

export const Appear = props => {
  const arr = React.Children.toArray(props.children)
  const step = useSteps(arr.length)
  const children = arr.map((child, i) =>
    i < step
      ? child
      : React.cloneElement(child, {
          style: {
            ...child.props.style,
            visibility: 'hidden',
          },
        })
  )

  return <>{children}</>
}

export default Appear
