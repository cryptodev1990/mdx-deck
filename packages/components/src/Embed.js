/** @jsx jsx */
// experimental component for embedding MDX Decks
// in other React applications
/*
 *
 *    import React from 'react'
 *    import { DeckEmbed } from '@mdx-deck/components'
 *    import deck from './my-deck.mdx'
 *
 *    export default props =>
 *      <>
 *        <h1>The first slide</h1>
 *        <DeckEmbed src={deck} />
 *        <h1>The second slide</h1>
 *        <DeckEmbed src={deck} slide={2} />
 *      </>
 *
 */

import { jsx } from '@emotion/core'
import React from 'react'
import { MDXProvider } from '@mdx-js/react'
import Provider from './Provider'
import Slide from './Slide'
import GoogleFonts from './GoogleFonts'

// from @mdx-deck/gatsby-theme
// todo: reorganize things
const splitter = props => {
  const { theme, themes } = props
  const arr = React.Children.toArray(props.children)
  const splits = []
  const slides = []

  arr.forEach((child, i) => {
    if (child.props.mdxType === 'hr') splits.push(i)
  })

  let previousSplit = 0
  splits.forEach(i => {
    const children = [...arr.slice(previousSplit, i)]
    slides.push(() => children)
    previousSplit = i + 1
  })
  slides.push(() => [...arr.slice(previousSplit)])

  return {
    ...props,
    theme,
    themes,
    slides,
  }
}

const Ratio = ({ ratio, children }) => (
  <div
    css={{
      outline: '1px solid tomato',
      width: '100%',
      height: 0,
      paddingBottom: ratio * 100 + '%',
      overflow: 'hidden',
      position: 'relative',
    }}
  >
    {children}
  </div>
)

const Placeholder = ({ index }) => (
  <pre style={{ fontSize: 16 }}>not found: slide {index}</pre>
)

const wrapper = props => {
  const { slides, theme, themes, ratio, zoom } = splitter(props)
  const Content = slides[props.slide - 1] || Placeholder

  return (
    <Provider theme={theme} themes={themes}>
      <GoogleFonts />
      <Ratio ratio={ratio}>
        <Slide
          style={{
            position: 'absolute',
            top: 0,
            bottom: 0,
            width: '100%',
            height: '100%',
            transform: `scale(${zoom})`,
          }}
        >
          <Content index={props.slide - 1} />
        </Slide>
      </Ratio>
    </Provider>
  )
}

export const Embed = ({ src: Deck, slide = 1, ratio = 9 / 16, zoom = 1 }) => (
  <MDXProvider
    components={{
      wrapper,
    }}
  >
    <Deck slide={slide} ratio={ratio} zoom={zoom} />
  </MDXProvider>
)

export default Embed
