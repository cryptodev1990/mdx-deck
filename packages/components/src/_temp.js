// may cut some of this

const GoogleFonts = withTheme(
  props =>
    !!props.theme.googleFont && (
      <link href={props.theme.googleFont} rel="stylesheet" />
    )
)
