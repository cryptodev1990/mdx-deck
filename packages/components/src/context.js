import React, { useContext } from 'react'

export const Context = React.createContext(null)

export const withContext = Component => props => (
  <Context.Consumer
    children={context => <Component {...props} context={context} />}
  />
)

export const useDeck = () => useContext(Context)
