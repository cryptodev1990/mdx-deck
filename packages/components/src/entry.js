import React from 'react'
import { render } from 'react-dom'
import { MDXDeck } from './index'

const mod = require(FILENAME)
const { slides } = mod
const { theme, components, Provider } = mod

console.log(slides)

export default class App extends React.Component {
  render() {
    return (
      <MDXDeck
        slides={slides}
        theme={theme}
        components={components}
        Provider={Provider}
      />
    )
  }
}

if (typeof document !== 'undefined') {
  render(<App />, document.getElementById('root'))
}

if (module.hot) module.hot.accept()
