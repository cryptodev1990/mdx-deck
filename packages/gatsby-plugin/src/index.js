import React from 'react'
import { MDXProvider } from '@mdx-js/react'
import { ThemeProvider } from 'theme-ui'
import wrapper from './deck'
import * as mdxComponents from './components'
import theme from './theme'

const components = {
  wrapper,
  ...mdxComponents,
}

const Page = props =>
  <MDXProvider components={components}>
    {props.children}
  </MDXProvider>

export const wrapPageElement = ({ element, props }) =>
  <Page {...props}>
    {element}
  </Page>

export {
  Notes,
  Head,
  Header,
  Footer,
  Image,
  Appear,
} from './components'
export { useDeck } from './context'
export { useSteps } from './hooks'
