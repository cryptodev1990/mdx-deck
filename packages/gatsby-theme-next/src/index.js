import React from 'react'
import App from './components/app'

export { Steps } from './components/steps'
export { Notes } from './components/notes'
export { Head } from './components/head'

export const wrapPageElement = ({ element }) => <App>{element}</App>
