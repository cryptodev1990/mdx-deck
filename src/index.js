/* v1 API
export { withDeck, withSlide } from './context'
export { Head } from './Head'
export { default as SlideDeck } from './SlideDeck'
export { default as Image } from './Image'
export { default as Notes } from './Notes'
export { default as Appear } from './Appear'
export { default as Code } from './Code'
export { default as components } from './components'

// themes
export { default as theme } from './themes'
export * as themes from './themes'

// internals for third-party components
export * as updaters from './updaters'
export * as constants from './constants'
*/

import React from 'react'
import PropTypes from 'prop-types'
import { Router, globalHistory, navigate } from '@reach/router'
import styled, { ThemeProvider } from 'styled-components'

export const Context = React.createContext(null)

export const withContext = Component => props => (
  <Context.Consumer
    children={context => <Component {...props} context={context} />}
  />
)

const themed = key => props => props.theme[key]

// TODO check against v1 styles
const SlideRoot = styled.div(
  {
    display: 'flex',
    flexDirection: 'column',
    width: '100vw',
    height: '100vh',
    alignItems: 'center',
    justifyContent: 'center',
  },
  themed('Slide')
)

const Slide = ({ children, ...props }) => (
  <Context.Provider value={props}>
    <SlideRoot>{children}</SlideRoot>
  </Context.Provider>
)

export class MDXDeck extends React.Component {
  constructor(props) {
    super(props)
    console.log(props.slides)

    this.state = {
      slides: props.slides,
      step: 0,
    }
  }

  handleKeyDown = ({ key }) => {
    switch (key) {
      case 'ArrowLeft':
        this.previous()
        break
      case 'ArrowRight':
        this.next()
        break
    }
  }

  getIndex = () => {
    const { pathname } = globalHistory.location
    return Number(pathname.split('/')[1] || 0)
  }

  getMeta = i => {
    const { slides } = this.state
    const { meta = {} } = slides[i] || {}
    return meta
  }

  goto = i => {
    const current = this.getIndex()
    const reverse = i < current
    navigate('/' + i)
    const meta = this.getMeta(i)
    this.setState({
      step: reverse ? meta.steps || 0 : 0,
    })
  }

  previous = () => {
    const { slides, step } = this.state
    const index = this.getIndex()
    const meta = this.getMeta(index)
    if (meta.steps && step > 0) {
      this.setState(state => ({
        step: state.step - 1,
      }))
    } else {
      const previous = index - 1
      if (previous < 0) return
      this.goto(previous)
    }
  }

  next = () => {
    const { slides, step } = this.state
    const index = this.getIndex()
    const meta = this.getMeta(index)
    if (meta.steps && step < meta.steps) {
      this.setState(state => ({
        step: state.step + 1,
      }))
    } else {
      const next = index + 1
      if (next > slides.length - 1) return
      this.goto(next)
    }
  }

  register = (index, meta) => {
    const { slides } = this.state
    slides[index].meta = meta
    this.setState({ slides })
  }

  componentDidMount() {
    document.body.addEventListener('keydown', this.handleKeyDown)
  }

  componentWillUnmount() {
    document.body.removeEventListener('keydown', this.handleKeyDown)
  }

  render() {
    const { slides } = this.state
    const context = {
      ...this.state,
      register: this.register,
    }

    const [FirstSlide] = slides

    return (
      <Router>
        <Slide path="/" index={0} {...context}>
          <FirstSlide path="/" />
        </Slide>
        {slides.map((Component, i) => (
          <Slide key={i} path={i + '/*'} index={i} {...context}>
            <Component path={i + '/*'} />
          </Slide>
        ))}
      </Router>
    )
  }
}

MDXDeck.propTypes = {
  slides: PropTypes.array.isRequired,
  theme: PropTypes.object.isRequired,
  components: PropTypes.object,
  Provider: PropTypes.func,
}
MDXDeck.defaultProps = {
  slides: [],
  theme: {},
}

// Additional API
// export const Appear
