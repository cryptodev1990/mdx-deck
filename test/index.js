import 'jest-styled-components'
import React from 'react'
import { create as render } from 'react-test-renderer'
import { MDXDeck } from '../src'

const renderJSON = el => render(el).toJSON()

describe('MDXDeck', () => {
  test('renders', () => {
    const json = renderJSON(<MDXDeck slides={[() => <h1>hi</h1>]} />)
    expect(json).toMatchSnapshot(``)
  })

  test.todo('handles keydown events')
})
