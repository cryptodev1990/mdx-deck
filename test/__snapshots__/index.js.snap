// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MDXDeck renders 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100vw;
  height: 100vh;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 {
  font-family: system-ui,sans-serif;
  color: #000;
  background-color: white;
  font-size: 1.5em;
  text-align: center;
}

.c0 li > ul,
.c0 li > ol {
  font-size: inherit;
}

.c0 li > p {
  font-size: inherit;
  margin: 0;
}

.c0 ul {
  text-align: left;
}

.c0 ol {
  text-align: left;
}

@media screen and (min-width:64em) {
  .c0 {
    font-size: 3em;
  }
}

<div
  onTouchStart={[Function]}
>
  <div
    className="c0"
  >
    <div
      role="group"
      style={
        Object {
          "outline": "none",
        }
      }
      tabIndex="-1"
    >
      <div
        className="c1"
      >
        <div
          role="group"
          style={
            Object {
              "outline": "none",
            }
          }
          tabIndex="-1"
        >
          <h1>
            hi
          </h1>
        </div>
      </div>
    </div>
  </div>
</div>
`;
